bootloader/piboot.go

```diff
@@ -0,0 +1,431 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootloader
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader/ubootenv"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// sanity - piboot implements the required interfaces
+var (
+	_ Bootloader                             = (*piboot)(nil)
+	_ ExtractedRecoveryKernelImageBootloader = (*piboot)(nil)
+	_ NotScriptableBootloader                = (*piboot)(nil)
+)
+
+const (
+	pibootCfgFilename = "piboot.conf"
+	pibootPartFolder  = "/piboot/ubuntu/"
+	// TODO The ubuntu-seed folder should be eventually passed
+	// around when creating the bootloader
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
+)
+
+func getSeedPartDirRuntime() string {
+	return ubuntuSeedDir
+}
+
+func getRebootParamPathRuntime() string {
+	return "/run/systemd/reboot-param"
+}
+
+// Get absolute dirs via variables so we can mock in tests
+var (
+	getSeedPartDir     = getSeedPartDirRuntime
+	getRebootParamPath = getRebootParamPathRuntime
+)
+
+type piboot struct {
+	rootdir string
+	basedir string
+}
+
+func (p *piboot) setDefaults() {
+	p.basedir = "/boot/piboot/"
```

Comment:
just a comment do we also need now extra mounts from the initramfs for this in: https://github.com/snapcore/core-initrd/blob/f783d1d9472e88581c88963b8c4efd1a35dcefcc/factory/usr/lib/the-modeenv#L11-L19

Reply:
We do need that, yes. Thanks for reminding me that I need to open a PR for core-initrd with that.